dialogue-core/src/main/java/com/palantir/dialogue/core/BalancedNodeSelectionStrategyChannel.java

```diff
@@ -189,48 +291,52 @@ public void onFailure(Throwable throwable) {
             return maybe;
         }
 
-        SortableChannel computeScore() {
-            int requestsInflight = inflight.get();
-            double failureReservoir = recentFailuresReservoir.get();
-
-            // it's important that scores are integers because if we kept the full double precision, then a single 4xx
-            // would end up influencing host selection long beyond its intended lifespan in the absence of other data.
-            int score = requestsInflight + Ints.saturatedCast(Math.round(failureReservoir));
-
-            observability.debugLogComputedScore(requestsInflight, failureReservoir, score);
-            return new SortableChannel(score, this);
+        ChannelStats getSnapshot() {
+            return new ChannelStats(this, inflight.get(), recentFailuresReservoir.get(), rtt.getNanos());
         }
 
         @Override
         public String toString() {
-            return "MutableChannelWithStats{score=" + computeScore().score
+            return "MutableChannelWithStats{"
+                    + "delegate=" + delegate
                     + ", inflight=" + inflight
                     + ", recentFailures=" + recentFailuresReservoir
-                    + ", delegate=" + delegate
+                    + ", rtt=" + rtt
                     + '}';
         }
     }
 
     /**
-     * A dedicated immutable class ensures safe sorting, as otherwise there's a risk that the inflight AtomicInteger
+     * A dedicated value class ensures safe sorting, as otherwise there's a risk that the inflight AtomicInteger
      * might change mid-sort, leading to undefined behaviour.
      */
-    private static final class SortableChannel {
-        private final int score;
+    @VisibleForTesting
+    static final class ChannelStats {
         private final MutableChannelWithStats delegate;
+        private final int requestsInflight;
+        private final double recentBadness;
+
+        private final long rttNanos;
+        private float rttSpectrum = 0; // assigned later once we've figured out the range of rttNanos across channels
```

Comment:
its a bit weird that we construct an instance of this and then later mutate the private field. Perhaps we can invert the flow and require a spectrum be provided when constructed?

Reply:
also its probably worth documenting (and checking) that this is a value between 0 and 1

Reply:
We can only compute this number when we have full information of all the rttNanos (cos we need to know the min and max across all channels), so we can't initialize this at the same times as the other fields.

If we introduced an extra type, maybe this would be clearer, i.e. ChannelSnapshot and ChannelSnapshot2?

Reply:
Right you could expose a way of getting `rtt.getNanos()` from each `MutableChannelWithStats` and then have `getSnapshot` take a `rttSpectrum` as a parameter.